File: microservice-architecture/distributed-transactions/ts/test/test_saga.ts
import { SagaCoordinator } from "../src/saga";

function assertEqual<T>(expected: T, actual: T, message: string): void {
  if (expected !== actual) {
    throw new Error(`${message} expected=${expected} actual=${actual}`);
  }
}

const success = new SagaCoordinator(10, false);
const completed = success.execute("ORD-1001", "SKU-BOOK", 2);
assertEqual("COMPLETED", completed.status, "successful saga");
assertEqual(8, success.inventory.bookStock, "stock should decrease");

const failure = new SagaCoordinator(10, true);
const cancelled = failure.execute("ORD-1002", "SKU-BOOK", 2);
assertEqual("CANCELLED", cancelled.status, "failing saga should cancel");
assertEqual(10, failure.inventory.bookStock, "stock should be compensated");

console.log("distributed-transactions(ts) tests passed");

File: microservice-architecture/distributed-transactions/python/test/test_saga.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.saga import SagaCoordinator  # noqa: E402


class DistributedTransactionsTest(unittest.TestCase):
    def test_saga_success(self) -> None:
        coordinator = SagaCoordinator(10, False)
        order = coordinator.execute("ORD-1001", "SKU-BOOK", 2)
        self.assertEqual("COMPLETED", order.status)
        self.assertEqual(8, coordinator.inventory.book_stock)

    def test_saga_compensation(self) -> None:
        coordinator = SagaCoordinator(10, True)
        order = coordinator.execute("ORD-1002", "SKU-BOOK", 2)
        self.assertEqual("CANCELLED", order.status)
        self.assertEqual(10, coordinator.inventory.book_stock)


if __name__ == "__main__":
    unittest.main()
